packages/controllers/src/snaps/SnapController.ts

```diff
@@ -1034,7 +1039,28 @@ export class SnapController extends BaseController<
       }
 
       this._recordSnapRpcRequest(snapName);
-      return handler(origin, request);
+
+      // handle max request time
+      let resolve: (...args: unknown[]) => void;
+      let reject: (...args: unknown[]) => void;
+
+      const timeoutPromise = new Promise((res, rej) => {
+        resolve = res;
+        reject = rej;
+      });
+
+      const timeout = setTimeout(() => {
+        this._stopSnap(snapName);
+        reject(new Error('request timed out'));
+      }, this._maxRequestTime);
```

Comment:
We can make this a bit more concise and avoid breaking out `resolve`/`reject`:

```suggestion
      // Handle max request time
      let timeout: number;

      const timeoutPromise = new Promise((_resolve, reject) => {
        timeout = setTimeout(() => {
          this._stopSnap(snapName);
          reject(new Error('The request timed out.'));
        }, this._maxRequestTime) as unknown as number;
      });
```